game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java

```diff
@@ -100,6 +102,11 @@ private void clearImageCache() {
     icons.clear();
   }
 
+  public Image getImage(final UnitCategory unit) {
+    return getImage(unit.getType(), unit.getOwner(), (unit.getDamaged() > 0), unit.getDisabled())
+        .orElseThrow(() -> new RuntimeException("No unit image for: " + unit));
```

Comment:
It's not an anti-pattern necessarily to throw RuntimeException, it depends on context. If it were, the language developers perhaps would have made the class abstract.

I tried really hard to find a great reference @ssoloff pointed me to, it was a guide by the google java people on which exceptions to throw when. RuntimeException was okay to throw when other specific exceptions were not appropriate and when to indicate a "developers messed up" type of situation.

In this case, IllegalState is not appropriate as the data state is fine, the file system has a file in an incorrect location or is completely missing. It's not quite the same as say "stream is already closed", "file has already been read". "IllegalArgumentException" is not appropriate because the argument is fine, the file system state is incorrect (IllegalState is perhaps arguably okay, but not necessarily an improvement).

IMO RuntimeExceptions should be subclassed if there is any expectation for the error to be handled, or if there is a good chance the error will be shown to a user. It is also useful if there is a lot of debugging to be attached. For example we could create a `new MissingImageException(File file)` that would be somewhat useful. That *is* a good thing to do IMO for map images, where map makers do have problems of placing image files in the correct place. On the other hand for errors where the devs messed up, and it's a one-off error that we'll always see simply by executing the code, the new exception I think will mostly be code clutter and not add that much value. The value being added is a more descriptive name to the error when the error message in this case makes it clear enough.

I was not able to find any blog posts that clearly stated that runtime exception in this type of situation is an anti-pattern. I do wish I could find that original reference from google on which exception to throw when, it was a great guide on this subject.

I found these two perhaps relevant posts/blogs discussing this topic:

1. https://dzone.com/articles/implementing-custom-exceptions-in-java See: "1. Always Provide a Benefit"
1. https://medium.com/@rafacdelnero/11-mistakes-java-developers-make-when-using-exceptions-af481a153397  See: "2 — Creating lots of specific Exceptions"

> Don’t create Exceptions for everything. Your application will be full of classes, useless repetition and you will create unnecessary work. Instead, create Exceptions for really important business requirements.

Reply:
> I tried really hard to find a great reference @ssoloff pointed me to, it was a guide by the google java people on which exceptions to throw when.

I think the reference you're looking for is [here](https://github.com/google/guava/wiki/ConditionalFailuresExplained).

Reply:
Indeed, thanks @ssoloff !

Do you have any thoughts on the above conversation?  Do you think my line of thinking holds water, or do you think it is always better to sub-class Runtime exception? Curious for the 3rd opinion here.

Reply:
> It's not an anti-pattern necessarily to throw RuntimeException, it depends on context. If it were, **the language developers perhaps would have made the class abstract.**

Bloch appears to agree that that should have been the case, but wasn't done to maintain backwards compatibility.  Per **Item 72: Favor the use of standard exceptions** in _Effective Java_, Third Edition (emphasis Bloch's):

> **Do _not_ reuse Exception, RuntimeException, Throwable, or Error directly.** Treat these classes as if they were abstract.

Personally, I typically try not to throw `RuntimeException` directly.  I agree that creating an excessive number of exception subclasses is probably not the best idea, but having a few generic application-specific subclasses that could be used in various scenarios to keep in the spirit of Item 72 might not be a bad idea.

However, in the present case, it seems you want to simply provide a custom exception message when the `Optional` is empty; otherwise, you would have just called `get()`.  If so, why not reuse the same exception `get()` would throw if the `Optional` were empty?  Namely, `NoSuchElementException`.  You don't plan on handling it, so it doesn't really matter what it is as long as it's not `RuntimeException`.